### PR TITLE
Fix #390 Issue Snapshot Plugin

### LINKED ISSUE #390: Snapshot tool
Create a tool to snapshot the map. 
As a user, when I click on the snapshot tool a window should show a preview of the snapshot with some information (date, size...) and allow me to download the image. 

The tool should work of course with OpenLayers and Leaflet maps. 

Here a mockup: 
snapshot tool

Here some useful link:
* Original Plugin in MapStore (1) : https://github.com/geosolutions-it/MapStore/blob/master/mapcomposer/app/static/externals/gxp/src/script/plugins/PrintSnapshot.js
* Server side part of the original Plugin: https://github.com/geosolutions-it/servicebox/blob/master/src/main/java/it/geosolutions/servicebox/UploadCanvas.java

### PR DESCRIPTION
**Pull request by kappu72**
Base implementation of snapshot components for ol and leaflet.
With ie or when image size is bigger then 2Mb, servicebox service is needed to allow the image saving.
In Leaflet html2canvas lib needed.
In Openlayers, it clone the map status, so any layer introduce form other component but not in map state will not be rendere.

### CODE REVIEW

File: web/client/components/map/openlayers/plugins/WMSLayer.js [hunk 1]
```diff
@@ -10,7 +10,9 @@ var Layers = require('../../../../utils/openlayers/Layers');
 var ol = require('openlayers');
 var objectAssign = require('object-assign');
 var CoordinatesUtils = require('../../../../utils/CoordinatesUtils');
-const {isArray} = require('lodash');
+var ConfigUtils = require('../../../../utils/ConfigUtils');
```
Comment (offtherailz) [inline web/client/components/map/openlayers/plugins/WMSLayer.js]: use const

File: web/client/components/map/openlayers/plugins/WMSLayer.js [hunk 2]
```diff
@@ -29,16 +31,63 @@ function getWMSURLs( urls ) {
     return urls.map((url) => url.split("\?")[0]);
 }
 
+function needProxy(uri) {
```
Comment (offtherailz) [inline web/client/components/map/openlayers/plugins/WMSLayer.js]: we should extenalize in a utils function. axios already have a function like this, isn't it?

File: web/client/examples/viewer/index.html [hunk 3]
```diff
@@ -25,6 +25,16 @@
             padding: 0;
             overflow: hidden;
         }
+        #snapshot_map{
```
Comment (offtherailz) [inline web/client/examples/viewer/index.html]: I think this should stay in the css of the component that initialize it, included with require. 
we should also use something like "snapshot_hidden_map"

File: web/client/components/Snapshot/SnapshotPanel.jsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            <Grid header={this.props.name} className="snapshot-panel" fluid={true}>
+                <Row>
+                    <Col xs={7} sm={7} md={7}><Image src={this.props.snapshot.img.data} responsive thumbnail/></Col>
+                    <Col xs={5} sm={5} md={5}>
+                       <Table responsive>
+                            <tbody>
+                               <tr>
+                                <td>Date</td><td> <DateFormat dateParams={this.props.dateFormat}/></td>
+                                </tr>
+                                <tr><td>Layers</td><td>{this.renderLayers()}</td></tr>
```
Comment (offtherailz) [inline web/client/components/Snapshot/SnapshotPanel.jsx]: comma and space between layers, to allow automatic new lines

**Comment by offtherailz** [comment]
The container window should be bigger

File: web/client/components/Snapshot/leaflet/GrabMap.jsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            this.props.onSnapshotReady(imgData, canvas.width, canvas.height, size);
+            this.props.onStatusChange("READY");
+        }
+    },
+    mapIsLoading(layers) {
+        return layers.some((layer) => layer.loading);
+    }
+});
+
+module.exports = GrabLMap;
```
Comment (offtherailz) [inline web/client/components/Snapshot/leaflet/GrabMap.jsx]: Leaflet preview has many issues. 
* Sometimes the map is not complete, layer missing or not well stacked 
image

* When swich layer or background, the preview do not update. 
image
* it is slow

File: web/client/components/Snapshot/SnapshotPanel.jsx [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            return (<Row style={{marginTop: "5px"}}>
+                    {this.props.snapshot.error}
+                    </Row>);
+        }
+    },
+    render() {
+        return ( (this.props.snapshot.img && this.props.snapshot.img.data) || this.props.snapshot.state === "READY") ? (
+            <Grid header={this.props.name} className="snapshot-panel" fluid={true}>
+                <Row>
+                    <Col xs={7} sm={7} md={7}><Image src={this.props.snapshot.img.data} responsive thumbnail/></Col>
```
Comment (offtherailz) [inline web/client/components/Snapshot/SnapshotPanel.jsx]: The thumbnail should show loading when the snapshot is updating.

File: web/client/components/Snapshot/SnapshotPanel.jsx [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                        </Table>
+                    </Col>
+                </Row>
+
+                <Row className="pull-right" style={{marginTop: "5px"}}>
+                    {(this.props.browser.ie || (this.props.snapshot.img.size > 2000 && this.props.browser.chrome)) ? this.remoteDownload() : this.localDownload()}
+                </Row>
+                {this.renderError()}
+            </Grid>
+        ) : null;
```
Comment (offtherailz) [inline web/client/components/Snapshot/SnapshotPanel.jsx]: On first render, the window should show loading messages, not a blank window
image

File: web/client/components/map/openlayers/plugins/WMSLayer.js (hunk 2 above)
Comment (kappu72) [inline web/client/components/map/openlayers/plugins/WMSLayer.js]: Axios has almost identical function but implemented in request interceptor.
Where you want to implement this function?

File: web/client/examples/viewer/index.html (hunk 3 above)
Comment (kappu72) [inline web/client/examples/viewer/index.html]: I agree

**Comment by offtherailz** [comment]
Summarizing:
* The Openlayers Snapshot is good
* The leaflet snapshot is buggy
* The snapshot preview need some improvement.

File: web/client/components/Snapshot/SnapshotPanel.jsx (hunk 4 above)
Comment (kappu72) [inline web/client/components/Snapshot/SnapshotPanel.jsx]: Changed to use ul li elements

**Comment by offtherailz** [comment]
Some error on preview creation. 
image

**Comment by offtherailz** [comment]
Cross origin is not used anymore for leaflet. 
image

**Comment by offtherailz** [comment]
During implementation the preview is very slow when using proxy.
Can we display a tainted canvas for the preview while generating the usable snapshot?

**Comment by offtherailz** [comment]
There is no warning about the background not supported (Google, Bing). 
When the snapshot is open with one of theese enabled, you should force one allowed background (e.g. openstreetmap) and write an info row on the snapshot preview window about it.

**Comment by offtherailz** [comment]
The snapshot window should be a little bigger, like in the previous comment

**Comment by offtherailz** [comment]
The loading is cool, but the window still remains blank for too much time, and it is confusing.

**Pull request closed without merging**